### PR TITLE
Fix bug 1243

### PR DESCRIPTION
**Pull request by gera-gas**
Resolves: #1243 

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"

Windows with error information for Temporary directory cases:
image

### CODE REVIEW

**Comment by JamesCrook** [comment]
We usually aim to have a help button on error dialogs to take you to the manual for more information.

**Comment by gera-gas** [comment]
> 
> 
> We usually aim to have a help button on error dialogs to take you to the manual for more information.

We have a link in this dialog to Audacity manual that followed to Directory Preferences description or you means, button used usual for this purpose (link to manual)? not a link.

**Comment by petersampsonaudacity** [comment]
An example
image

**Comment by JamesCrook** [comment]
I like the (new) use of the html window in an error dialog as it gives easy option of hyperlink, cut-and-paste-ability and wrapping that does not depend on the translators.

Notice that Peter's example also has an informative title for the dialog.

**Comment by DavidBailes** [comment]
> I like the (new) use of the html window in an error dialog as it gives easy option of hyperlink

Unfortunately, the hyperlink won't be accessible to keyboard users.

**Comment by crsib** [comment]
The error dialog is changed to: https://crsib-screenshots.s3.amazonaws.com/20210715_18.29.58_gaaab.png

File: src/widgets/UnwritableLocationErrorDialog.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#pragma once
+
+	
+#include "widgets/wxPanelWrapper.h"
+
+/// A error dialog about unwritable location, that allows to navigate to settings quickly
```
Comment (Paul-Licameli) [inline src/widgets/UnwritableLocationErrorDialog.h:16]: *An

File: src/widgets/UnwritableLocationErrorDialog.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#include "ui/AccessibleLinksFormatter.h"
+
+BEGIN_EVENT_TABLE(UnwritableLocationErrorDialog, wxDialogWrapper)
+    EVT_BUTTON(wxID_OK, UnwritableLocationErrorDialog::OnOk)
+    EVT_BUTTON(wxID_HELP, UnwritableLocationErrorDialog::OnError)
+END_EVENT_TABLE()
+
+UnwritableLocationErrorDialog::UnwritableLocationErrorDialog(wxWindow* parent, const wxString& path)
+    : wxDialogWrapper(
+         parent, -1, XO("Error"), wxDefaultPosition, wxDefaultSize, wxCAPTION | wxCLOSE_BOX)
```
Comment (Paul-Licameli) [inline src/widgets/UnwritableLocationErrorDialog.cpp:26]: Is it important to exclude wxSYSTEM_MENU from the style, then?

File: src/widgets/UnwritableLocationErrorDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+         S.StartVerticalLay();
+         {
+            S.AddFixedText(
+               /* i18n-hint: %s is replaced with a directory path. */
+               XO("Unable to write files to directory: %s.").Format(path),
+               false, 500);
+
+            S.AddFixedText(
+               /* i18n-hint: This message describes the error in the Error dialog. */
+               XO("Please check the directory exists, has the necessary permissions and the drive isn't full."),
```
Comment (Paul-Licameli) [inline src/widgets/UnwritableLocationErrorDialog.cpp:49]: "Please check that the directory exists, has the necessary permissions, and the drive isn't full."

"that" and one "Oxford comma' improve the English a bit.

File: src/widgets/UnwritableLocationErrorDialog.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+               false, 0);
+
+            S.AddSpace(0, 8);
+
+            AccessibleLinksFormatter preferencesMessage(
+               /* i18n-hint: %s is replaced with 'Preferences > Directories'. */
+               XO("You can change the directory in %s."));
+
+            preferencesMessage.FormatLink(
+               /* i18n-hint: Hyprelink title that opens Preferences dialog on Directories page. */
```
Comment (Paul-Licameli) [inline src/widgets/UnwritableLocationErrorDialog.cpp:59]: *hyperlink

File: src/widgets/UnwritableLocationErrorDialog.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+               wxT("%s"), XO("Preferences > Directories"), [parent, this]() {
+                  EndModal(wxID_OK);
+
+                  GlobalPrefsDialog dialog(parent, nullptr);
+
+                  dialog.SelectPageByName(XO("Directories").Translation());
```
Comment (Paul-Licameli) [inline src/widgets/UnwritableLocationErrorDialog.cpp:65]: I'd like to change SelectPageByName to take const TranslatableString&.  Not blocking this PR.

File: src/widgets/UnwritableLocationErrorDialog.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+void UnwritableLocationErrorDialog::OnOk(wxCommandEvent&)
+{
+   EndModal(wxID_OK);
+}
+
+void UnwritableLocationErrorDialog::OnError(wxCommandEvent&)
+{
+   HelpSystem::ShowHelp(this, "Error:_Disk_full_or_not_writable", false);
+}
```
Comment (Paul-Licameli) [inline src/widgets/UnwritableLocationErrorDialog.cpp:102]: This dialog class looks good, aside form minor string and comment changes, but I'm suspecting we will use FormatLink with callback much more often for similar help to the user, and should think about generalizing this dialog into a more easily reused class.  For later.

File: src/prefs/DirectoriesPrefs.cpp [hunk 7]
```diff
@@ -418,6 +418,11 @@ bool DirectoriesPrefs::Validate()
    }
    else {
       /* If the directory already exists, make sure it is writable */
+      if (!FileNames::WritableLocationCheck(mTempText->GetValue()) ||
+          !FileNames::WritableLocationCheck(mMacrosText->GetValue()))
+      {
+          return false;
+      }
```
Comment (Paul-Licameli) [inline src/prefs/DirectoriesPrefs.cpp:425]: I understand, this will prevent closing of the Preferences dialog, so long as either path is bad.

**Review by Paul-Licameli — APPROVED** [review]

**Pull request merged**